File: widgets/data_input_card.py
"""
单个数据输入卡片组件。
紧凑单行布局：状态指示灯 + 标签 + 输入框 + 复制按钮。
"""

from PyQt5.QtCore import Qt, pyqtSignal
from PyQt5.QtWidgets import (
    QWidget, QHBoxLayout, QLabel, QLineEdit, QPushButton, QApplication,
    QSizePolicy,
)

import resources as res


class DataInputCard(QWidget):
    """数据输入卡片：标签和输入框在同一行的紧凑控件"""

    value_changed = pyqtSignal(str)

    def __init__(self, card_id, label_text, hint_text, parent=None):
        super().__init__(parent)
        self._card_id = card_id

        self.setSizePolicy(QSizePolicy.Expanding, QSizePolicy.Fixed)
        self.setStyleSheet(f"""
            DataInputCard {{
                background-color: {res.BG_CARD};
                border-radius: 4px;
            }}
        """)

        layout = QHBoxLayout(self)
        layout.setContentsMargins(8, 4, 8, 4)
        layout.setSpacing(6)

        # 状态指示灯
        self._dot = QLabel()
        self._dot.setFixedSize(6, 6)
        self._dot.setStyleSheet(f"""
            background-color: {res.BORDER_DEFAULT};
            border-radius: 3px;
            border: none;
        """)
        layout.addWidget(self._dot, alignment=Qt.AlignVCenter)

        # 标签
        self._label = QLabel(label_text)
        self._label.setFixedWidth(130)
        self._label.setStyleSheet(
            f"font-size: {res.FONT_BODY}px; font-weight: 600; "
            f"color: {res.TEXT_TITLE}; border: none;"
        )
        layout.addWidget(self._label, alignment=Qt.AlignVCenter)

        # 输入框
        self._input = QLineEdit()
        self._input.setPlaceholderText(hint_text)
        self._input.setStyleSheet(res.input_style())
        self._input.textChanged.connect(self._on_text_changed)
        layout.addWidget(self._input, 1)

        # 复制按钮
        btn_copy = QPushButton("复制")
        btn_copy.setFixedSize(40, 24)
        btn_copy.setStyleSheet(f"""
            QPushButton {{
                background-color: {res.BG_MAIN};
                color: {res.PRIMARY};
                border: 1px solid {res.PRIMARY};
                border-radius: 4px;
                font-size: {res.FONT_HINT}px;
            }}
            QPushButton:hover {{
                background-color: {res.PRIMARY_LIGHT};
            }}
        """)
        btn_copy.clicked.connect(self._on_copy)
        layout.addWidget(btn_copy, alignment=Qt.AlignVCenter)

    def _on_text_changed(self, text):
        self.set_status("ok" if text.strip() else "empty")
        self.value_changed.emit(text)

    def _on_copy(self):
        QApplication.clipboard().setText(self._input.text())

    def set_value(self, text):
        self._input.blockSignals(True)
        self._input.setText(text)
        self._input.blockSignals(False)

    def get_value(self):
        return self._input.text().strip()

    def set_status(self, status):
        colors = {"empty": res.BORDER_DEFAULT, "ok": res.SUCCESS, "error": res.ERROR}
        self._dot.setStyleSheet(f"""
            background-color: {colors.get(status, res.BORDER_DEFAULT)};
            border-radius: 3px;
            border: none;
        """)
        if status == "error":
            self._input.setStyleSheet(res.input_error_style())
        else:
            self._input.setStyleSheet(res.input_style())

    def set_visible(self, visible):
        self.setVisible(visible)

    def card_id(self):
        return self._card_id

File: src/ui/DataInputCard.cpp
#include "ui/DataInputCard.h"
#include "ui/Theme.h"
#include <QHBoxLayout>
#include <QApplication>
#include <QClipboard>
#include <algorithm>

DataInputCard::DataInputCard(const QString& cardId, const QString& label,
                             const QString& hint, QWidget* parent)
    : QWidget(parent)
    , m_cardId(cardId)
    , m_hint(hint)
{
    setMinimumHeight(32);
    setStyleSheet(QStringLiteral(
        "background-color: %1; border-radius: %2px;")
        .arg(Theme::BG_CARD).arg(Theme::BORDER_RADIUS));

    auto* layout = new QHBoxLayout(this);
    layout->setContentsMargins(12, 4, 12, 4);
    layout->setSpacing(8);
    layout->setAlignment(Qt::AlignVCenter);

    // Status dot
    m_statusDot = new QLabel(this);
    m_statusDot->setFixedSize(8, 8);
    m_statusDot->setStyleSheet(QStringLiteral(
        "background-color: %1; border-radius: 4px;").arg(Theme::TEXT_HINT));
    layout->addWidget(m_statusDot);

    // Label — centered, adaptive width
    m_label = new QLabel(label, this);
    m_label->setMinimumWidth(90);
    m_label->setMaximumWidth(140);
    m_label->setAlignment(Qt::AlignVCenter);
    m_label->setStyleSheet(QStringLiteral(
        "font-size: %1px; font-weight: 600; color: %2; border: none;")
        .arg(Theme::FONT_BODY).arg(Theme::TEXT_TITLE));
    layout->addWidget(m_label);

    // Input — adaptive font via resizeEvent
    m_input = new QLineEdit(this);
    m_input->setPlaceholderText(hint);
    m_input->setStyleSheet(Theme::inputStyle());
    layout->addWidget(m_input, 1);

    // Copy button
    m_btnCopy = new QPushButton(QStringLiteral("复制"), this);
    m_btnCopy->setFixedSize(56, 30);
    m_btnCopy->setStyleSheet(QStringLiteral(R"(
        QPushButton {
            background-color: #FFFFFF;
            color: %1;
            border: 1px solid %1;
            border-radius: 4px;
            font-size: %2px;
        }
        QPushButton:hover { background-color: %3; }
    )").arg(Theme::PRIMARY).arg(Theme::FONT_HINT).arg(Theme::PRIMARY_LIGHT));
    layout->addWidget(m_btnCopy);

    connect(m_input, &QLineEdit::textChanged, this, &DataInputCard::onTextChanged);
    connect(m_btnCopy, &QPushButton::clicked, this, &DataInputCard::onCopy);
}

QString DataInputCard::cardId() const { return m_cardId; }

QString DataInputCard::value() const { return m_input->text().trimmed(); }

void DataInputCard::setValue(const QString& text)
{
    m_input->blockSignals(true);
    m_input->setText(text);
    m_input->blockSignals(false);
    setStatus(text.trimmed().isEmpty() ? QStringLiteral("empty") : QStringLiteral("ok"));
}

void DataInputCard::setStatus(const QString& status)
{
    QString color;
    if (status == "ok")
        color = Theme::SUCCESS;
    else if (status == "error")
        color = Theme::ERROR;
    else
        color = Theme::TEXT_HINT;

    m_statusDot->setStyleSheet(QStringLiteral(
        "background-color: %1; border-radius: 4px;").arg(color));

    if (status == "error") {
        m_input->setStyleSheet(Theme::inputErrorStyle());
    } else {
        m_input->setStyleSheet(Theme::inputStyle());
    }
}

void DataInputCard::onTextChanged(const QString& text)
{
    QString status = text.trimmed().isEmpty() ? "empty" : "ok";
    setStatus(status);
    emit valueChanged(text.trimmed());
}

void DataInputCard::resizeEvent(QResizeEvent* event)
{
    QWidget::resizeEvent(event);
    int fs = std::clamp(height() / 3, 11, 14);
    m_input->setStyleSheet(Theme::inputStyle(fs));
}

void DataInputCard::onCopy()
{
    QApplication::clipboard()->setText(m_input->text().trimmed());
}
